modules/rest-compatibility/src/main/java/org/elasticsearch/compat/TypeConsumer.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.compat;
+
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.rest.RestRequest;
+
+import java.util.HashSet;
+import java.util.Set;
+import java.util.function.Function;
+
+public class TypeConsumer implements Function<String, Boolean> {
```

Comment:
I think this class has a couple problems... 
it is named Consumer, but actually a Function with a side effect
it requires an ordering between methods. (e.g. if you call hasTypes can yield different results then calling apply then hasTypes)

`hasTypes()` method feels utility like, and `apply` is a pure function (which shouldn't have side effects even local to it's own class).

Can we separate this into two classes , a utility class and pure function class. If you need to hold to the state returned by the function, whoever calls that function can take that responsibility. 

Also, in tangent with my other comment on renaming the variable, I think the function class be be generalized as well `KeyAllowed` ?

Reply:
Agree - the name is not right. The consumer was an initial idea but it ended up a function and I didn't change it accordingly. The KeyAllowed also feels right to me.

I am not sure I understand how to refactor this. My idea was that the `typeConsumer` object being created at the V7 layer and passed down to v8 code can bubble up the state (information if type was found) back to v7 layer. Allowing the code to take action (most of the time just deprecation warning) to be present in v7 layer.

The `apply()` method was meant to serve two functions. Retrieve and store the information for v7 layer if the type was found. And to tell v8 code that the key was allowed and exception when parsing is not necessary.

I also agree that side effect is probably unexpected. The `TypeConsumer` is visible as just a `Function` in v8 layer, so noone looking just at code which uses it would expect side effects.
If we have two classes, that would mean we would need to pass down two objects down to v8 layer?

Reply:
@jakelandis I think this is the main issue on that PR.
We can split this in two functions 
- KeyAllowed<String,Boolean> - a function returning true if a string is of allowed value (_type, types etc) false otherwise. This will then be pass down into a Request parsing logic - i.e. https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/action/termvectors/TermVectorsRequest.java#L588
- second class - not sure about this one.. we need to go down from v7 rest layer to [the same place](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/action/termvectors/TermVectorsRequest.java#L588) where a request is parsed (v8 code). 
I understand that it is not obvious that a function would has a state. But from the parsing point of view it wouldn't matter. All it care in TermVectorsRequest.java#L588 is if the field is allowed. The state if the field occurred (type for that example) is only relevant in v7 code where a class can be used directly (not by the Function interface)

If we were to have a utility class it would have live within some common to all modules utility module. That would not be ideal I guess.
We could refactor hasTypes to be a utility that is consuming the function that was passed down. But I am not sure this would benefit much (except from solving the ordering)

Anyways I think I might be missing something in your idea of splitting this class into two. Can you please give more details?